### PR TITLE
bug 1981498: add vulnerable legacy injector to allow for upgrade clusters to use ...

### PR DESCRIPTION
**Pull request by deads2k**
the old service-ca.crt content in SA tokens

This is part 1 of a four part plan.

1. add the ability to inject *only* an openshift-service-ca.crt configmap with content that in 4.7 was contained in token SA service-ca.crt.  It uses the `service.alpha.openshift.io/inject-vulnerable-legacy-cabundle` annotation so that it is clearly undesireable and the behavior isn't guaranteed.  The controller here yields to the preferred controller if both annotations are set.
2. add an API field to the kcm operator resource so a user can indicate they wish to switch to a more secure service-ca.crt.
3. Add a flag to the KCM to indicate "use the vulnerable annotation" during configmap creation. The bootstrap kcm does not need this since it is only used in new clusters.
4. set the new API field into the kcm-o "create-only" operator resource.  This will let new clusters start secure.  The operator will use the field to set the KCM flag

### CODE REVIEW

**Comment by deads2k** [comment]
the lack of unit tests is going to be sticky. :(

I guess we'll use the e2e-aws run.

```
=== RUN   TestE2E/vulnerable-legacy-ca-bundle-injection-configmap
    e2e_test.go:1376: configmaps "openshift-service-ca.crt" already exists
```

**Comment by mfojtik** [comment]
/lgtm


/hold
(for Standa ack)

File: pkg/controller/cabundleinjector/starter.go [hunk 1]
```diff
@@ -44,21 +52,36 @@ func StartCABundleInjector(ctx context.Context, controllerContext *controllercmd
...
 	// depending on the operator for redeployment
 	caBundleFile := "/var/run/configmaps/signing-cabundle/ca-bundle.crt"
-
 	caBundleContent, err := ioutil.ReadFile(caBundleFile)
 	if err != nil {
 		return err
 	}
 
+	// this construction matches what the old kube controller manager did. It added the entire ca.crt to the service-ca.crt.
+	vulnerableLegacyCABundleContent, err := ioutil.ReadFile(caBundleFile)
```
Comment (stlaz) [inline pkg/controller/cabundleinjector/starter.go:61]: nit: this re-reads the file

**Comment by stlaz** [comment]
/hold cancel
looks ok

**Comment by deads2k** [comment]
reapplying lgtm

File: test/e2e/e2e_test.go [hunk 2]
```diff
@@ -1317,6 +1338,90 @@ func TestE2E(t *testing.T) {
...
+			if err != nil {
+				t.Fatal(err)
+			}
+		}
+
+		// the content should now change pretty quick.  We sleep because it's easier than writing a new poll and I'm pressed for time
+		time.Sleep(5 * time.Second)
+		publishedConfigMap, err = adminClient.CoreV1().ConfigMaps(ns.Name).Get(context.TODO(), publishedConfigMap.Name, metav1.GetOptions{})
+
+		// if we changed the injection, we should see different content
```
Comment (marun) [inline test/e2e/e2e_test.go:1413]: nit: Maybe move this comment to the second block or remove it entirely? It doesn't appear relevant to the contents of the first block, since in that case different content is not expected.

**Review by marun — COMMENTED** [review]
Some nits...
